### PR TITLE
Clip snap to time

### PR DESCRIPTION
**Pull request by vsverchinsky**
Resolves: #4285 

Also fixes wrong guidline positions when clips are moved between tracks

Currently adds to new commits on top of #4414

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/tracks/labeltrack/ui/LabelTrackShifter.cpp [hunk 1]
```diff
@@ -134,12 +134,14 @@ class LabelTrackShifter final : public TrackShifter {
...
 
-   bool Attach( Intervals intervals ) override
+   bool Attach( Intervals intervals, double offset ) override
    {
       auto pTrack = mpTrack.get();
       std::for_each( intervals.rbegin(), intervals.rend(),
-         [this, pTrack](auto &interval){
+         [this, pTrack, offset](auto &interval){
             auto pData = static_cast<IntervalData*>( interval.Extra() );
+            if(offset == .0)
```
Comment (Paul-Licameli) [inline src/tracks/labeltrack/ui/LabelTrackShifter.cpp:143]: Do you mean `!-`?

File: src/tracks/ui/TimeShiftHandle.cpp [hunk 2]
```diff
@@ -227,7 +227,7 @@ bool TrackShifter::AdjustFit(
    return false;
 }
 
-bool TrackShifter::Attach( Intervals )
+bool TrackShifter::Attach( Intervals, double offset )
```
Comment (Paul-Licameli) [inline src/tracks/ui/TimeShiftHandle.cpp:230]: Might omit `offset`

File: src/tracks/labeltrack/ui/LabelTrackShifter.cpp (hunk 1 above)
Comment (vsverchinsky) [reply src/tracks/labeltrack/ui/LabelTrackShifter.cpp:143]: Yes, thanks!

File: libraries/lib-wave-track/WaveTrack.cpp
Comment (Paul-Licameli) [inline libraries/lib-wave-track/WaveTrack.cpp:1]: In commit comment “guidelines”

File: src/tracks/ui/TimeShiftHandle.cpp [hunk 3]
```diff
@@ -565,71 +565,59 @@ UIHandle::Result TimeShiftHandle::Click
...
-            if (pInterval) {
-               clipLeft = pInterval->Start() + desiredSlideAmount;
-               clipRight = pInterval->End() + desiredSlideAmount;
-            }
-            else {
-               clipLeft = capturedTrack.GetStartTime() + desiredSlideAmount;
-               clipRight = capturedTrack.GetEndTime() + desiredSlideAmount;
-            }
+      auto track = state.mCapturedTrack.get();
+      double clipLeft, clipRight;
```
Comment (Paul-Licameli) [inline src/tracks/ui/TimeShiftHandle.cpp:576]: Initializers please

File: src/tracks/playabletrack/wavetrack/ui/WaveTrackShifter.cpp [hunk 4]
```diff
@@ -153,15 +153,24 @@ class WaveTrackShifter final : public TrackShifter {
...
          auto pData = static_cast<WaveTrack::IntervalData*>( interval.Extra() );
          auto pClip = pData->GetClip();
          if ( !mpTrack->AddClip( pClip ) )
             return false;
          mMigrated.insert( pClip.get() );
-         mMoving.emplace_back( std::move( interval ) );
+         if(offset == .0)
+            mMoving.emplace_back( std::move( interval ) );
+         else
+         {
```
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveTrackShifter.cpp:167]: Were the side effects on interval caused by move in the first branch important?  This branch omits them

Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveTrackShifter.cpp:167]: I suppose not, now I see they belong to expiring `remover`

**Comment by Paul-Licameli** [comment]
The older version adjusts `desiredSlideAmount` first to quantize, then for snapping, then for AdjustFit (small tolerances allowed by WaveClip, but depending also on zoom level -- time per pixel column), then for `AdjustOffsetSmaller`.

The new version adjusts `desiredSlideAmount` for quantization, then for AdjustFit, then for snapping, then for `AdjustOffsetSmaller`.

That was difficult to figure out, in all this logic I had forgotten about for long.

Maybe this makes a subtle change of behavior of clip moving in some cases?  But maybe this does not matter.

**Comment by Paul-Licameli** [comment]
> The older version adjusts `desiredSlideAmount` first to quantize, then for snapping, then for AdjustFit (small tolerances allowed by WaveClip, but depending also on zoom level -- time per pixel column), then for `AdjustOffsetSmaller`.
> 
> The new version adjusts `desiredSlideAmount` for quantization, then for AdjustFit, then for snapping, then for `AdjustOffsetSmaller`.
> 
> That was difficult to figure out, in all this logic I had forgotten about for long.
> 
> Maybe this makes a subtle change of behavior of clip moving in some cases? But maybe this does not matter.

I'm thinking it through more carefully.

In the old version, I think`AdjustOffsetSmaller` does not matter in case of moving between tracks, because all the other checks for fit already adjust the offset, so it will do nothing.  Only in case of sliding horizonally within track, does this matter.  `AdjustFit` does all the protection against overlapping clips.

But with this change, snapping may make another change of horizontal position of clips after those checks!

But then, `AdjustOffsetSmaller` really does work to prevent overlap.

So really it's just the composition of a correct vertical move and a correct horizontal move.

We won't get overlapping clips.  We might devise some test case where a drag between clips is permitted in one version but not the other.  Because the check for fit might give different answers at the snapped and unsnapped x coordinates.

**Comment by Paul-Licameli** [comment]
I tried this case:  make three clips in sequence that touch exactly.  Drag the middle one to another empty track below.  Then try to drag it back again into the gap that fits it exactly.

Actually it's hard to do in both versions, in neither one, just working simply.  It seems in the old version I must drag horizontally until snapping, then move vertically without any more lateral movement.

I'm not sure, maybe the new makes it slightly easier but still it doesn't always just work.

Maybe you can figure out an extra fix for this case.

But, as for exploring consequences, to find any reason to dealy this PR -- no, I haven't found one.

**Review by Paul-Licameli — APPROVED** [review]

**Comment by Paul-Licameli** [comment]
I would just ask for some more comments in `Drag` making explicit what I said about the correctness of clip fit in all case.

**Pull request merged**